hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java

```diff
@@ -0,0 +1,1061 @@
+      if (dirNSSummary == null) {
+        if (withReplica) {
+          duResponse.setSizeWithReplica(0L);
+        }
+        break;
+      }
+
+      Set<Long> subdirs = dirNSSummary.getChildDir();
+
+      duResponse.setKeySize(dirNSSummary.getSizeOfFiles());
+      long dirDataSize = duResponse.getKeySize();
+      long dirDataSizeWithReplica = 0L;
+      List<DUResponse.DiskUsage> subdirDUData = new ArrayList<>();
+      // iterate all subdirectories to get disk usage data
+      for (long subdirObjectId: subdirs) {
+        NSSummary subdirNSSummary =
+            reconNamespaceSummaryManager.getNSSummary(subdirObjectId);
+        String subdirName = subdirNSSummary.getDirName();
+        // build the path for subdirectory
+        String subpath = buildSubpath(normalizedPath, subdirName);
+        DUResponse.DiskUsage diskUsage = new DUResponse.DiskUsage();
+        // reformat the response
+        diskUsage.setSubpath(subpath);
+        long dataSize = getTotalSize(subdirObjectId);
+        dirDataSize += dataSize;
+
+        if (withReplica) {
+          long subdirDU = calculateDUUnderObject(subdirObjectId);
+          diskUsage.setSizeWithReplica(subdirDU);
+          dirDataSizeWithReplica += subdirDU;
+        }
+
+        diskUsage.setSize(dataSize);
+        subdirDUData.add(diskUsage);
+      }
+
+      // handle direct keys under directory
+      if (listFile || withReplica) {
+        dirDataSizeWithReplica += handleDirectKeys(dirObjectId, withReplica,
+            listFile, subdirDUData, normalizedPath);
+      }
+
+      if (withReplica) {
+        duResponse.setSizeWithReplica(dirDataSizeWithReplica);
+      }
+      duResponse.setCount(subdirDUData.size());
+      duResponse.setSize(dirDataSize);
+      duResponse.setDuData(subdirDUData);
+      break;
+    case KEY:
+      // DU for key doesn't have subpaths
+      duResponse.setCount(0);
+      // The object ID for the directory that the key is directly in
+      long parentObjectId = getDirObjectId(names, names.length - 1);
+      String fileName = names[names.length - 1];
+      String ozoneKey =
+          omMetadataManager.getOzonePathKey(parentObjectId, fileName);
+      OmKeyInfo keyInfo =
+          omMetadataManager.getFileTable().getSkipCache(ozoneKey);
+      duResponse.setSize(keyInfo.getDataSize());
+      if (withReplica) {
+        long keySizeWithReplica = getKeySizeWithReplication(keyInfo);
+        duResponse.setSizeWithReplica(keySizeWithReplica);
+      }
+      break;
+    case UNKNOWN:
+      duResponse.setStatus(ResponseStatus.PATH_NOT_FOUND);
+      break;
+    default:
+      break;
+    }
+    return Response.ok(duResponse).build();
+  }
+
+  /**
+   * Quota usage endpoint that summarize the quota allowed and quota used in
+   * bytes.
+   * @param path request path
+   * @return Quota Usage response
+   * @throws IOException
+   */
+  @GET
+  @Path("/quota")
+  public Response getQuotaUsage(@QueryParam("path") String path)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.